### DISCUSSION TITLE
Defining GC State in ODF

### DISCUSSION THREAD

**Discussion by jefferycrowley** [GrandOrgue/grandorgue#1930] [Help]
Hello,

I have an issue of GO ODF not placing anything in the ODF at all if you define the GC State as No Change, On or OFF.   I have tried many times to program either a stop, noise, or switch to NOT touch the item if GC is called.   For instance the blower and noise should stay in whatever state that it was prior to the GC being used.   

Also I am having a rough time with GO ODF - I am using 0.12.2 version, and somehow it is leaving off all of the Percussive=N out of the ODF unless you have decided to change the default to =N and Grand Orgue should just read the ODF without an issue.   But I go back and double-check all stops and ranks as percussive=N and then re-write the ODF with GO ODF.   When I attempt to launch the compiled ODF it gives me an error and hard stop upon loading with the message "missing necessary element - percussive on Stop XX or Rank XX"  When I look back in the ODF via Notepad - indeed only if I have said "Y" to percussive is it in the ODF.   Otherwise it is not placed in the description of the stop or rank.    

I have resorted to saving the ODF that I have fixed correctly, re-naming the GO ODF to a different name and then manually copying the correct entry that needs to be changed.  Am I doing something incorrectly?   AND.. how should the text in the ODF look like?

GCState=No Change
GCState=Off
GCState=On?

or is there another syntax to use?

**Comment by ClarkeBrunt** [comment]
From the GO help (for Drawstop) GCState "(integer -1 to 1, default: implementation defined) State of the button after pressing GC. -1 means no change, 0 off and 1 on.".

I don't know about GOODF. Unless the behaviour of GCState has been altered since my previous observations in the discussions, then not supplying GCState at all (therefore, supposedly, behaviour implementation defined) seems in practice to mean 'turn off'.

**Reply by larspalo** [reply]
The GCState would only be used if the drawstop itself has a function of Input (since only then it's actually possible to use it independently). To not have a drawstop affected by the GC the correct setting in GoOdf is "No Change" and in the text odf it's GCState=-1. Again, if the function of a drawstop is something else than Input - then GC state is actually controlled by some other switch and its GCState setting.

As noted in GoOdf README.md file there are changes in GO about the Percussive option that in turn affects the backwards compatibility of odfs edited/created with GoOdf:

> Please be aware that currently, due to changes in the GrandOrgue software, from the version 0.11.0 of GoOdf the .organ files written will require a GrandOrgue version 3.14+ to work! If compatibility with older versions of GO is important one must also use an older version of GoOdf (like 0.10.0) to write the odf.

The solution is to use recent versions of GO together with recent versions of GoOdf. It would have been rather trivial to implement it so that odfs would be written backwards compatible in regards to the Percussive option, but that would mean writing out currently completely un-necessary lines - which is something I really try to avoid at all costs (it's a main design goal of GoOdf, that I've set for myself, to try to only keep writing that which is absolutely necessary and remove all that which is anyway automatically handled/calculated by GO). Besides, the current state might be an incentive to make users keep their GO installation up to date...

**Comment by jefferycrowley** [comment]
This is another quick question.   Thanks for the feedback on the GC State - and I did have a new GOODF and the last version of GO so I updated the GO version to latest, but still I have issues with the item "percussive=N" causing hard stops when loading an new ODF and had to put in the item line in pretty-much everything to get the ODF to open up.   I even deleted the last version of GO.   The error happens in my Organ Computer which has the same exact version OF GO.   I have GOODF on my Office PC and the latest version of GO on the Office PC as well.   I have checked the dates on everything and size as well.  They all match up exactly, But, the office PC gives no errors or hard stops when "percussive=N" is missing and does not need to be there at all.

My extra question is regards to having a written explanation for the GOODF GUI additional choices as "Available Setter Elements" and starts with [CrescendoA...  Most of the list I am familiar with how to implement the feature, but there are several that I have not been able to figure out such as "CurrFileName - which I believe should be the name of the organ? or the *.organ name.   There does not seem to be a good way to configure it.   I tried to add a "LCD" screen on the Main Panel of the organ that I am creating, there is no clue as to where the "Name" either implies or how to configure the element.   Also the elements "GeneralXX" or "SetterXXXDivisionalYYY" , PrevBank, NextBank are self explainatory, but GOODF only allows for a Stop Head (or pistion type) to be used instead of an "LCD Screen that has the particular Setter position, the General just used, Previous or Next Bank.  It seemingly does not use any text with the exception of the supposed stepper frame, but when I tested it out the digits in the stop head did not change when I added a new frame to the stepper.  My understanding is that the element feature should indeed have the stepper frame AND the GENERAL last used.  Same for the other similar buttons.

And what is [L0, L1,L2] etc.  I assume they are Memory Levels and you would have to put a button for every level needed instead of having a single button/image that would invoke the Memory Leveler and an up and down arrow button or image to scroll up or down the memory banks, which would only require a companion "label" or "LCD Screen" that would show the current memory level on it?   Perhaps for the future - but by doing it this way there only has to be 4 images on the Panel instead of 12 items.

I did get the Temperament element to work correctly by adding an "LCD Screen and two arrows up and down" to scroll through the temperament choices correctly.

And for my real Organ Console how do you configure 2x16 LCD screens to work with Grand Orgue.   I have tried every way I know to configure the Organ Name, the Temperament, the Setter Frame, the General last used, the Memory Level.   I have all of these set-up in Hauptwerk and it only took a few minutes to set-up Globally for every organ and they work flawlessly.   The main LCD screen also shows the loading progress before changing to the Organ Name and Temperament.   The Setter LCD also shows the Memory Level current and the name I have given the Combination group and it loads immediately upon changing the registration group.   

I currently have 3 - 2x16 characters LCD screens.   You can also set them up to show the percentage of the expression pedal as well, but I have 3 Bargraph LED lights and One 3 Digit with decimal readout for the crescendo pedal.

Keep up the great work with both GO and GOODF.   GOODF takes just about as much time as doing it from scratch - but it is easier to configure due to the ODF just simply connecting all of the disparate parts of an organ definition.  The graphic editor is brilliant and takes a lot of the guess work out of the whole GUI stuff.   I create my organ images in GIMP and get it set 100% as three separate screens for the panels.   I can just copy the attributes of the item into GOODF and I rarely have to move anything around in Grand Orgue with GOODF, but it makes quick work if everything is done before hand with another Graphic editor.

**Reply by larspalo** [reply]
If you have a GO version of 3.14+ the Percussive=N is optional as it's the default value inherited from the [Organ] and down - thus it's not necessary to specify anymore unless it's overriding a parent that have Percussive=Y set.

Most of the SetterElements are the same as are used on the GO built in panels. They offer an easy way to include the same elements on any panel. You'll have to read the GO help, GoOdf help and experiment to learn all about them. It's much too much to write in a single post here.

GoOdf show you what type of element each setter element is by giving you the appropriate attributes to change, when you select any one added to a panel. Only label derived elements as such can have a "LCD" sort of appearance/functionality. So, you'll need to look for the different setter element labels to combine the information available. Also, they present their information as the name attribute so you cannot manually set it!

> "CurrFileName - which I believe should be the name of the organ? or the *.organ name.

No, it's the currently loaded settings/combinations file that is active, as presented on the sequencer or "Combination Setter" panel.

> And what is [L0, L1,L2] etc. I assume they are Memory Levels

They are the single digit number buttons as presented on the sequencer or "Combination Setter" panel with __0, __1 etc.

> And for my real Organ Console how do you configure 2x16 LCD screens to work with Grand Orgue.

You'll have to configure elements (from panels) to send MIDI information to the LCD screens. Right click an element, configure its MIDI send events appropriately.

In File->Settings->Initial MIDI the elements that can be MIDI pre-configured for any organ exist. If you find a need for more elements like the setter element labels for combination numbers, temperaments, current combination file etc to be available for global configuration I suggest you create an issue as a feature request for them, otherwise they must be set for each organ.
